### ISSUE TITLE
[3.0.0-beta1] tests fail for debug build of solaris64-sparcv9-cc

### ISSUE THREAD

**Issue by makr** [openssl/openssl#15978]
Testing on a Solaris 11.3 sun4v sparc machine fails for the `--debug` build. On the same machine the tests of the `--release` build are all `ok`, however. Compiler is a ...
```
$ cc -V
cc: Studio 12.5 Sun C 5.14 SunOS_sparc 2016/05/31
```

[Configure-dump.txt](https://github.com/openssl/openssl/files/6752819/Configure-dump.txt)

Tests are ok until ...
```
05-test_pbe.t ......................
    # ERROR: (bool) 'PKCS5_PBE_keyivgen(ctx, pbe_password, strlen(pbe_password), algor->parameter, cipher, md, 1) == true' failed @ test/pbetest.c:76
    # false
    # 00000001:error:0308010C:digital envelope routines:(unknown function):unsupported:crypto/evp/evp_fetch.c:346:Global default library context, Algorithm (PBKDF1 : 0), Properties (<null>)
    # OPENSSL_TEST_RAND_ORDER=1625209038
    not ok 1 - test_pkcs5_pbe_rc4_md5
# ------------------------------------------------------------------------------
    # ERROR: (bool) 'PKCS5_PBE_keyivgen(ctx, pbe_password, strlen(pbe_password), algor->parameter, cipher, md, 1) == true' failed @ test/pbetest.c:76
    # false
    # 00000001:error:0308010C:digital envelope routines:(unknown function):unsupported:crypto/evp/evp_fetch.c:346:Global default library context, Algorithm (PBKDF1 : 0), Properties (<null>)
    # OPENSSL_TEST_RAND_ORDER=1625209038
    not ok 2 - test_pkcs5_pbe_des_sha1
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/pbetest => 1
not ok 1 - Running PBE test
05-test_pbe.t ...................... 1/? ---------------------------------------
#   Failed test 'Running PBE test'
#   at test/recipes/05-test_pbe.t line 27.
05-test_pbe.t ...................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
```
... then again ok, until ...
```
20-test_enc.t ...................... 1/?
aes-128-cbc: unable to load provider legacy
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(../../providers/legacy.so): ld.so.1: openssl: fatal: relocation error: file ../../providers/legacy.so: symbol bn_expand2: referenced symbol not found
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_lib.c:162:
00000001:error:078C0105:common libcrypto routines:(unknown function):init fail:crypto/provider_core.c:658:name=legacy
../../util/wrap.pl ../../apps/openssl aes-128-cbc -bufsize 113 -e -k test -provider-path ../../providers -provider default -provider legacy -in ./p -out ./p.aes-128-cbc.cipher => 1
not ok 3 - aes-128-cbc
# ------------------------------------------------------------------------------
#   Failed test 'aes-128-cbc'
#   at test/recipes/20-test_enc.t line 70.
```
... all tests are failing now with similar messages, until ...
```
#   Failed test 'sm4-ofb'
#   at test/recipes/20-test_enc.t line 70.
sm4-ofb: unable to load provider legacy
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(../../providers/legacy.so): ld.so.1: openssl: fatal: relocation error: file ../../providers/legacy.so: symbol bn_expand2: referenced symbol not found
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_lib.c:162:
00000001:error:078C0105:common libcrypto routines:(unknown function):init fail:crypto/provider_core.c:658:name=legacy
../../util/wrap.pl ../../apps/openssl sm4-ofb -bufsize 113 -a -e -k test -provider-path ../../providers -provider default -provider legacy -in ./p -out ./p.sm4-ofb.cipher => 1
not ok 96 - sm4-ofb base64
# ------------------------------------------------------------------------------
#   Failed test 'sm4-ofb base64'
#   at test/recipes/20-test_enc.t line 70.
20-test_enc.t ...................... Dubious, test returned 94 (wstat 24064, 0x5e00)
Failed 94/96 subtests
20-test_enc_more.t ................. 1/?
enc: unable to load provider legacy
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(../../providers/legacy.so): ld.so.1: openssl: fatal: relocation error: file ../../providers/legacy.so: symbol bn_expand2: referenced symbol not found
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_lib.c:162:
00000001:error:078C0105:common libcrypto routines:(unknown function):init fail:crypto/provider_core.c:658:name=legacy
../../util/wrap.pl ../../apps/openssl enc -aes-128-cbc -k test -provider-path ../../providers -provider default -provider legacy -e -in ./testdatafile -out ./testdatafile.aes-128-cbc.cipher => 1
not ok 3 - aes-128-cbc
# ------------------------------------------------------------------------------
```
... all following tests fail now with similar messages, until ...
```
#   Failed test 'sm4-ecb'
#   at test/recipes/20-test_enc_more.t line 60.
enc: unable to load provider legacy
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(../../providers/legacy.so): ld.so.1: openssl: fatal: relocation error: file ../../providers/legacy.so: symbol bn_expand2: referenced symbol not found
00000001:error:12800067:DSO support routines:(unknown function):could not load the shared library:crypto/dso/dso_lib.c:162:
00000001:error:078C0105:common libcrypto routines:(unknown function):init fail:crypto/provider_core.c:658:name=legacy
../../util/wrap.pl ../../apps/openssl enc -sm4-ofb -k test -provider-path ../../providers -provider default -provider legacy -e -in ./testdatafile -out ./testdatafile.sm4-ofb.cipher => 1
not ok 137 - sm4-ofb
# ------------------------------------------------------------------------------
#   Failed test 'sm4-ofb'
#   at test/recipes/20-test_enc_more.t line 60.
20-test_enc_more.t ................. Dubious, test returned 135 (wstat 34560, 0x8700)
Failed 135/137 subtests
20-test_kdf.t ...................... ok
```
... there are quite a number of further tests failing, all of them with the complain that legacy.so cannot be loaded because of the missing symbol bn_expand2.

The summary:
```
Test Summary Report
-------------------
05-test_pbe.t                    (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
20-test_enc.t                    (Wstat: 24064 Tests: 96 Failed: 94)
  Failed tests:  3-96
  Non-zero exit status: 94
20-test_enc_more.t               (Wstat: 34560 Tests: 137 Failed: 135)
  Failed tests:  3-137
  Non-zero exit status: 135
30-test_evp.t                    (Wstat: 16384 Tests: 67 Failed: 64)
  Failed tests:  1-64
  Non-zero exit status: 64
80-test_cms.t                    (Wstat: 256 Tests: 12 Failed: 1)
  Failed test:  4
  Non-zero exit status: 1
80-test_pkcs12.t                 (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  7
  Non-zero exit status: 1
90-test_store.t                  (Wstat: 512 Tests: 8 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 416 tests but ran 8.
Files=239, Tests=2812, 1078 wallclock secs (11.94 usr  2.71 sys + 782.03 cusr 144.82 csys = 941.50 CPU)
Result: FAIL
```

Let me know, if you need further information.

**Comment by paulidale** [comment]
You are correct that the underlying problem here is the inability to load the legacy provider.  I suspect the differences are likely to be environmental.  Is there a difference in how you build the debug version?  Do you install either before testing?

**Comment by t8m** [comment]
Potential fix in #16019, please test.

**Comment by makr** [comment]
Sorry, couldn't make it in time. Have to wait for a window next week.
